source/server/hot_restarting_base.cc

```diff
@@ -0,0 +1,213 @@
+    const uint64_t cur_chunk_size = std::min(MaxSendmsgSize, total_size - sent);
+    iovec iov[1];
+    iov[0].iov_base = next_byte_to_send;
+    iov[0].iov_len = cur_chunk_size;
+    next_byte_to_send += cur_chunk_size;
+    sent += cur_chunk_size;
+    msghdr message;
+    memset(&message, 0, sizeof(message));
+    message.msg_name = &address;
+    message.msg_namelen = sizeof(address);
+    message.msg_iov = iov;
+    message.msg_iovlen = 1;
+
+    // Control data stuff, only relevant for the fd passing done with PassListenSocketReply.
+    uint8_t control_buffer[CMSG_SPACE(sizeof(int))];
+    if (replyIsExpectedType(&proto, HotRestartMessage::Reply::kPassListenSocket) &&
+        proto.reply().pass_listen_socket().fd() != -1) {
+      memset(control_buffer, 0, CMSG_SPACE(sizeof(int)));
+      message.msg_control = control_buffer;
+      message.msg_controllen = CMSG_SPACE(sizeof(int));
+      cmsghdr* control_message = CMSG_FIRSTHDR(&message);
+      control_message->cmsg_level = SOL_SOCKET;
+      control_message->cmsg_type = SCM_RIGHTS;
+      control_message->cmsg_len = CMSG_LEN(sizeof(int));
+      *reinterpret_cast<int*>(CMSG_DATA(control_message)) = proto.reply().pass_listen_socket().fd();
+      ASSERT(sent == total_size, "an fd passing message was too long for one sendmsg().");
+    }
+
+    const int rc = sendmsg(my_domain_socket_, &message, 0);
+    RELEASE_ASSERT(rc == static_cast<int>(cur_chunk_size),
+                   fmt::format("hot restart sendmsg() failed: returned {}, errno {}", rc, errno));
+  }
+  RELEASE_ASSERT(fcntl(my_domain_socket_, F_SETFL, O_NONBLOCK) != -1,
+                 fmt::format("Set domain socket nonblocking failed, errno = {}", errno));
+}
+
+bool HotRestartingBase::replyIsExpectedType(const HotRestartMessage* proto,
+                                            HotRestartMessage::Reply::ReplyCase oneof_type) {
+  return proto != nullptr && proto->requestreply_case() == HotRestartMessage::kReply &&
+         proto->reply().reply_case() == oneof_type;
+}
+
+// Pull the cloned fd, if present, out of the control data and write it into the
+// PassListenSocketReply proto; the higher level code will see a listening fd that Just Works. We
+// should only get control data in a PassListenSocketReply, it should only be the fd passing type,
+// and there should only be one at a time. Crash on any other control data.
+void HotRestartingBase::getPassedFdIfPresent(HotRestartMessage* out, msghdr* message) {
+  cmsghdr* cmsg = CMSG_FIRSTHDR(message);
+  if (cmsg != nullptr) {
+    RELEASE_ASSERT(cmsg->cmsg_level == SOL_SOCKET && cmsg->cmsg_type == SCM_RIGHTS &&
+                       replyIsExpectedType(out, HotRestartMessage::Reply::kPassListenSocket),
+                   "recvmsg() came with control data when the message's purpose was not to pass a "
+                   "file descriptor.");
+
+    out->mutable_reply()->mutable_pass_listen_socket()->set_fd(
+        *reinterpret_cast<int*>(CMSG_DATA(cmsg)));
```

Comment:
is there a way to sanity check this value?

Reply:
Well, you could check for negatives, but I'm not sure what the right response is. Crash, throw an exception, or just log a warning? Although the fd passing code is now basically unrecognizable, it's ultimately just a refactor of what was already there; I think the higher level logic is handing negatives (along with real looking but invalid ones).